# Домашнее задание №3

## Глобальные параметры работы блокнота

In [2]:
# Зафиксируем seed для воспроизводимости
seed = 1

# Выбранная контрольная точка модели
model_checkpoint = "google-t5/t5-small"
vocab_size = 50000

# # Очистка папки с моделью
! rm -rf ./Model/

## Настройка среды

Импорт необходимых библиотек:

In [3]:
import torch
import numpy as np
import pandas as pd

from transformers import AutoModel, AutoConfig, AutoTokenizer, AutoModelForSeq2SeqLM, T5ForConditionalGeneration, Seq2SeqTrainingArguments, DataCollatorForSeq2Seq, Seq2SeqTrainer
from datasets import load_dataset, Features, Value, concatenate_datasets, Dataset, DatasetDict
from huggingface_hub import notebook_login
from evaluate import load


Аутентификация в HuggingFace:

In [4]:
notebook_login() # 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

In [5]:
device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
print('{} device is available'.format(device))

if torch.cuda.is_available() and device.type == 'cuda':
    np.random.seed(seed)
     # Если используете CUDA, устанавливаем seed для GPU                
    torch.cuda.manual_seed_all(seed)
    # Для полной воспроизводимости также можно зафиксировать алгоритмы cuDNN
    torch.backends.cudnn.deterministic = True   
    torch.backends.cudnn.benchmark = False
else:
    # Устанавливаем seed для CPU генератора случайных чисел PyTorch
    np.random.seed(seed)
    torch.manual_seed(seed)


# Несколько оптимизаций которые могут ускорить процесс обучения
if torch.backends.cudnn.is_available():
    print(f'Using cudnn version: {torch.backends.cudnn.version()}')
    if device.type == 'cuda':
        torch.backends.cudnn.enabled = True
        torch.backends.cudnn.benchmark = True

cuda device is available
Using cudnn version: 90100


## Задание

[Ссылка](https://tn-gvu.mckx.ru/c/c4YcAACAurcBAEAA/sE43BQ/ZYcLPOYT_IPW6LK3/?u=https%3A%2F%2Fcontest.yandex.ru%2Fcontest%2F67637%2Fenter%2F%3Futm_source%3Dmindbox%26utm_medium%3Demail%26utm_campaign%3Dtraining6%26utm_content%3Ddigest%26utm_term%3D06.11.2024) на контест с заданием.

Необходимо обучить модель перевода с языка зетан на английский.

***Данные:*** https://disk.yandex.ru/d/u8mmcUBn64p_Nw

***Формат решения:*** json-lines в формате, аналогичной валидации, с переводами исходных текстов в тестовой выборке

***Метрика качества:*** BLEU между переводами и английскими референсами на закрытом тесте

## Загрузка данных

In [ ]:
# ! wget https://disk.yandex.ru/d/u8mmcUBn64p_Nw

In [6]:
data_files = {
    "train": "train.jsonl",
    "validation": "val.jsonl",
    "test": "test_no_reference.jsonl"
}

features = Features({
    'src': Value('string'),
    'dst': Value('string'),
})

dataset = load_dataset("json", data_files=data_files, features=features)
dataset

Generating train split: 0 examples [00:00, ? examples/s]

Generating validation split: 0 examples [00:00, ? examples/s]

Generating test split: 0 examples [00:00, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['src', 'dst'],
        num_rows: 300000
    })
    validation: Dataset({
        features: ['src', 'dst'],
        num_rows: 500
    })
    test: Dataset({
        features: ['src', 'dst'],
        num_rows: 1000
    })
})

Посмотрим на один пример данных из каждой части датасета:

In [7]:
dataset['train'][4]

{'src': '◈◠ ◧▱◠▦ ◀◫◓ ▨◠◉ ◂▱◠▽◈◠▦ ◀◠▷◞◪◈◗◳◧◓■ ◉◧◐▾▦▱◨◐▾ ○▱◎◠▦▱◠◓◈◠▦ ▨◠◉◠▦ ▽◠▷◨◈◫▱▴◓▵',
 'dst': "He's talking about a few right here in Lisbon, mostly Jews who have escaped the Germans, that's all."}

In [ ]:
dataset['validation'][]


{'src': '◲◒▱▴▫◗▱◪▦ ▻▾◀ ◚◪ ◀◠◓▱◠◓◈◠ ◀◨ ◠◳ ◳◫◳▴▼◪▨ ◞◠▫◬◒▱◠◓▪ ▽◭▢◈◪ ◭◉ ◈▩◒▴◓▨◪▦ ◗◉▨◫ ◞◠▫◬◒▱◠◓▪ ◳▩▢◈▴ 6■6 ◠◓▫▫▪▵"',
 'dst': 'Sales of drinks in pubs and bars increased by 6.6% over the month, while sales of food increased by 3%."'}

Немного узнать об этом чудном языке не помешает. Посмотрим сколько уникальных символов в этом языке:

In [9]:
# Функция для подсчета уникальных символов
def count_unique_characters(dataset, feature_name):
    unique_characters = set()
    for split in dataset.keys():
        for example in dataset[split][feature_name]:
            if example is not None:
                unique_characters.update(example)
    return len(unique_characters), unique_characters

# Подсчет уникальных символов во всех частях датасета
num_unique_chars, unique_chars = count_unique_characters(dataset, "src")

print(f"Количество уникальных символов: {num_unique_chars}")
print(f"Уникальные символы: {unique_chars}")

Количество уникальных символов: 182
Уникальные символы: {'Ý', 'ţ', 'É', '◓', 'ι', '$', 'Ä', '◜', '◈', '▰', '◄', '£', '◇', '◫', 'å', '◝', '1', ' ', '▲', '◰', '◢', '4', '_', '&', '◦', '\u200b', '◖', '◀', '▣', '▬', '◲', '◃', ')', '‚', 'ú', '▯', '+', 'Þ', '◩', '¿', '◭', '◑', '◮', 'ă', '◅', '^', 'ë', 'ä', '`', 'ø', '~', 'í', 'ß', '◂', '◚', 'º', ';', '◬', '◎', '◗', 'û', 'ο', '◨', '{', '▶', '/', '¤', '◡', '\x9e', '%', 'þ', '◛', '▨', '◧', '▦', '♪', '\\', '▾', '◞', '◙', '9', 'đ', 'ñ', '◒', 'ν', 'ª', '▼', '?', '7', '0', '■', '◍', '◯', '}', 'ï', '◌', '▵', '◆', '◟', '▹', '◳', 'Â', '▥', ']', '”', '–', '(', '▿', '▧', '´', '◥', '▭', '°', '▢', 'ó', '!', '5', 'é', '▤', '•', '§', '◱', 'î', '‘', 'ô', '─', '2', '“', '◊', 'â', '#', '½', '△', '◤', '◉', 'Î', '□', '*', '3', '◘', '8', '▴', '◁', '▮', '=', '▷', 'ý', '\u202d', '●', '\xa0', '6', 'è', '@', '◠', '◔', '™', '◣', '—', "'", 'á', '▪', ':', '\x99', '◕', '▩', 'ð', '¡', '◐', '►', '◪', '▸', '♫', '▻', '○', '▽', '[', '"', '\x9d', '¶', '▫', '’', '▱'}


Посмотрим аналогичную статистику для целевых последовательностей:

In [10]:
# Подсчет уникальных символов во всех частях датасета для целевых последовательностей
num_unique_chars_dst, unique_chars_dst = count_unique_characters(dataset, "dst")

print(f"Количество уникальных символов в целевых последовательностях: {num_unique_chars_dst}")
print(f"Уникальные символы в целевых последовательностях: {unique_chars_dst}")

Количество уникальных символов в целевых последовательностях: 201
Уникальные символы в целевых последовательностях: {'Ý', 'É', '˜', '$', 'Ä', 'Ð', '.', 'i', '£', 'å', 'N', 'Á', 'ş', '1', ' ', '¬', 'Q', '鈥', 'D', '4', '_', '&', 'œ', '│', 'υ', 'ò', 'у', 'ú', ')', '‚', '+', 't', ',', 'G', '¿', '‒', 'x', 'Ι', '©', 'ă', '¯', '^', 'ê', 'u', 'Æ', 'İ', 'ë', '`', 'ä', '±', '€', 'ø', 'í', '~', '\x92', ';', 'ί', 'ο', '{', 'p', '/', 'M', 'X', 'ü', '¤', 'þ', '%', 'V', 'Ã', 'T', 'S', 'A', 'm', 'Β', '♪', 'Ż', '\\', 'W', 'š', 'E', 'B', 'æ', '9', 'ñ', 'ν', 'ì', 'F', 'ª', 'Η', 'ē', '7', '?', 'õ', 'I', '0', '檛', 'fl', 'f', 'Ü', 'l', 'n', '}', 'ï', '\x8d', 'Z', 's', 'Â', 'w', 'ç', ']', '”', '-', 'b', 'ı', '–', '(', '´', 'h', 'O', 'k', 'v', 'Y', '°', '\xad', 'ó', 'g', 'L', '!', '5', 'é', '¢', 'Ö', '§', 'à', 'ρ', 'z', 'ô', '─', '2', '“', 'fi', 'Τ', 'с', 'â', 'e', '#', '½', '³', 'K', 'J', '*', 'Μ', 'Ν', '3', 'a', 'y', '8', 'ć', 'Ë', 'j', 'd', '=', 'C', 'ý', '\u202d', '\xa0', '●', 'c', 'U', '6', 'è', '@', '™', 

Посмотрим на максимальную длинну строк для каждой части датасета:

In [11]:
max_lengths = {}

for split in dataset:
    src_lengths = [len(item["src"]) for item in dataset[split] if item["src"] is not None]
    dst_lengths = [len(item["dst"]) for item in dataset[split] if item["dst"] is not None]
    
    max_src_length = max(src_lengths) if src_lengths else 0
    max_dst_length = max(dst_lengths) if dst_lengths else 0
    
    max_lengths[split] = {"src": max_src_length, "dst": max_dst_length}

print(max_lengths)

{'train': {'src': 321, 'dst': 395}, 'validation': {'src': 504, 'dst': 352}, 'test': {'src': 458, 'dst': 0}}


Самая длинная строка присутствующая в датасете имеет длинну 504 символа.

### Очистка данных

После разбора домашнего задания выяснелось, что данные были зашумлены. Вот некоторые из признаков проблемных переводов:
1) Длинна перевода больше длинны оргинала более чем в $n$ раз (и наоборот) Где $n$ определяется эмперически для каждого датасета;
2) В паре примера одно из полей (`src` и/или `dst`) пустое;
3) В переводе и оригинале используются арабские цифры, но они различаются;
4) в паре используются скобки и кавычки, они различаются.

Конвертируем обучающую часть датасета в Pandas Dataframe:

In [60]:
train_df = dataset['train'].to_pandas()
train_df.head()

,src,dst
0,◄▴◓◠▨ ◨▽◠▦◈◬◓▪▼◬▵,- Intriguing.
1,▽◪◎◗▦◫▦◫ ▫▴▨◓◠◓ ▴▫◎◪▱◫ ◚▴ ◞◧▦◞▾▢▱◨▨ ◒◠◓◠◀▪▦◈◠▦...,He would need to repeat his vows in the land o...
2,◄▴◞◠▸▱◠◓▪◎◠ ◀◫▱◪ ▼◪◚◠▻ ◚▴◓▴◎◪◈◗▦ ◎◫?,You couldn't even answer my texts?
3,▯◪ ▨◠◈◠◓ ◞◭◓◠▫ ◳◠▻◬◳◧◓ ◞▴▦◗▦▨◫?,How fast do you go?
4,◈◠ ◧▱◠▦ ◀◫◓ ▨◠◉ ◂▱◠▽◈◠▦ ◀◠▷◞◪◈◗◳◧◓■ ◉◧◐▾▦▱◨◐▾ ...,"He's talking about a few right here in Lisbon,..."


#### Непечатные символов

Проверим строки на наличие непечатных символов:

In [61]:
# Функция для поиска непечатных символов в строке
def find_non_printable(text):
    return re.search(r'[^\x20-\x7E]', text) is not None

# Найти все строки в признаках `src` и `dst`, которые содержат непечатные символы
non_printable_rows = train_df[train_df['src'].apply(find_non_printable) | train_df['dst'].apply(find_non_printable)]

# Вывести найденные строки
non_printable_rows

,src,dst
0,◄▴◓◠▨ ◨▽◠▦◈◬◓▪▼◬▵,- Intriguing.
1,▽◪◎◗▦◫▦◫ ▫▴▨◓◠◓ ▴▫◎◪▱◫ ◚▴ ◞◧▦◞▾▢▱◨▨ ◒◠◓◠◀▪▦◈◠▦...,He would need to repeat his vows in the land o...
2,◄▴◞◠▸▱◠◓▪◎◠ ◀◫▱◪ ▼◪◚◠▻ ◚▴◓▴◎◪◈◗▦ ◎◫?,You couldn't even answer my texts?
3,▯◪ ▨◠◈◠◓ ◞◭◓◠▫ ◳◠▻◬◳◧◓ ◞▴▦◗▦▨◫?,How fast do you go?
4,◈◠ ◧▱◠▦ ◀◫◓ ▨◠◉ ◂▱◠▽◈◠▦ ◀◠▷◞◪◈◗◳◧◓■ ◉◧◐▾▦▱◨◐▾ ...,"He's talking about a few right here in Lisbon,..."
...,...,...
299995,►◅▴▨◫◓◕▴ ▨◠▱◠◀◠▱◬◐▪▦ ◈▪◒◬▦◠ ◈◧◐◓▾ ▷◠◓▴▨▴▫ ▴◈◗◳◂◓▵,=Grasshopper is moving out of the crowd.=
299996,◝◓◳▼▴ ◉◂▼▾◐▾◎◠ ◫◳◫ ◀◠▨,"Brice, I need you take care of my child. I wil..."
299997,◅◂▨ ◒◗◓◫▦ ◀◗◓ ◉◂▼◨▨▫◨▵,He was such a sweet kid.
299998,◮▢◨▦ ▢◠◎◠▦◈▪◓ ◨▢◠▽◈◠ ◈▴◐◗▱◈◗◎▵,I haven't been in space for a long time.


In [58]:
# Функция для удаления непечатных символов из строки
def remove_non_printable(text):
    return re.sub(r'[^\x20-\x7E]', '', text)

# Применение функции к признакам 'src' и 'dst' в DataFrame
train_df['src'] = train_df['src'].apply(remove_non_printable)
train_df['dst'] = train_df['dst'].apply(remove_non_printable)

# Проверка результата
train_df.head()

,src,dst
0,,- Intriguing.
1,,He would need to repeat his vows in the land o...
2,?,You couldn't even answer my texts?
3,?,How fast do you go?
4,,"He's talking about a few right here in Lisbon,..."


#### Значения `None` в признаках `src` и `dst`

Найдём все строки в признаках которых встречается значение `None`:

In [55]:
import re

# Функция для поиска непечатных символов в строке
def find_non_printable(text):
    return re.search(r'[^\x20-\x7E]', text) is not None

# Найти все строки в признаках `src` и `dst`, которые содержат непечатные символы
non_printable_rows = train_df[train_df['src'].apply(find_non_printable) | train_df['dst'].apply(find_non_printable)]

# Вывести найденные строки
non_printable_rows

,src,dst,src_punctuation,dst_punctuation
1,▽◪◎◗▦◫▦◫ ▫▴▨◓◠◓ ▴▫◎◪▱◫ ◚▴ ◞◧▦◞▾▢▱◨▨ ◒◠◓◠◀▪▦◈◠▦...,He would need to repeat his vows in the land o...,{},{.}
2,◄▴◞◠▸▱◠◓▪◎◠ ◀◫▱◪ ▼◪◚◠▻ ◚▴◓▴◎◪◈◗▦ ◎◫?,You couldn't even answer my texts?,{?},"{?, '}"
3,▯◪ ▨◠◈◠◓ ◞◭◓◠▫ ◳◠▻◬◳◧◓ ◞▴▦◗▦▨◫?,How fast do you go?,{?},{?}
4,◈◠ ◧▱◠▦ ◀◫◓ ▨◠◉ ◂▱◠▽◈◠▦ ◀◠▷◞◪◈◗◳◧◓■ ◉◧◐▾▦▱◨◐▾ ...,"He's talking about a few right here in Lisbon,...",{},"{,, ., '}"
5,◝▴▦ ◫◒▫◪▦ ▨◠◳▫◠◓◠▼◠▨ ◀◗◓◫ ◈▴◐◫▱◈◗◓▵,"For instance, Ben is not a guy who skips work.",{},"{,, .}"
...,...,...,...,...
299995,►◅▴▨◫◓◕▴ ▨◠▱◠◀◠▱◬◐▪▦ ◈▪◒◬▦◠ ◈◧◐◓▾ ▷◠◓▴▨▴▫ ▴◈◗◳◂◓▵,=Grasshopper is moving out of the crowd.=,{},"{=, .}"
299996,◝◓◳▼▴ ◉◂▼▾◐▾◎◠ ◫◳◫ ◀◠▨,"Brice, I need you take care of my child. I wil...",{},"{,, ., '}"
299997,◅◂▨ ◒◗◓◫▦ ◀◗◓ ◉◂▼◨▨▫◨▵,He was such a sweet kid.,{},{.}
299998,◮▢◨▦ ▢◠◎◠▦◈▪◓ ◨▢◠▽◈◠ ◈▴◐◗▱◈◗◎▵,I haven't been in space for a long time.,{},"{., '}"


### Прямая речь

Сразу видно, что в признаке `dst` встречается прямая речь. Проверим это:

In [34]:
# Найти все строки в признаке `dst`, которые начинаются с `- `
dst_starts_with_dash = train_df[train_df['dst'].str.startswith('- ')]

# Вывести найденные строки
dst_starts_with_dash.head(10)

,src,dst
0,◄▴◓◠▨ ◨▽◠▦◈◬◓▪▼◬▵,- Intriguing.
22,◝▾ ◫◒◫ ◀◫◓ ◒◪▨◫▱◈◪▵▵▵ ► ▭◗◉ ◒◠▦◞◬▦ ◳◧▨▵,- We gotta work something...
36,◆◇▫◭◓◪▼◪◐◫◎■ ◝◪◀◠▼◬◐◬◎▵,"- Of course, Beba."
43,► ◝◠▦◠ ◳◠◓◈◬◎ ◪▫◎◪▦◗ ◗◞▫▴◎◫▽◂◓◨◎!,- I don't want you to help me!
47,► ◢◇◒◪◳◫ ◈◣▦▩▦▼▴ ◕◠▱◗◀◠▵,"- (Irene moans) - (Johnny) Come on, y'all."
55,► ◊◳◗ ▽◂▱▼▾▱◨▨▱◠◓▵,"- Have a ""bon voyage"""
94,◢◠▻◠ ◉◪▦◪▦◫ ◈▴ ◉▪▨◠◓ ▩◞▫◭▦◈▴▨◫▦◗▵,- Shut up and get outta your gear. - I'm takin...
113,► ▭◗▦◈◗ ▻◗◒◗◓▴◀◫▱◗◓ ◎◗◳◫◎?,- Can I make a turkey?
131,► ◱◓◠▱◠◓▪ ◈◫▦▱◪◎▴▨ ◫◞▫▴◓◗◎▵,- I want to hear about it.
152,◝◨◓◠◈◠ ▫▴▨ ◀◠◒▪◎▪▢◠ ◂▱◎◠◎▪▢ ◀▴▦◫ ◓◠▷◠▫◞▪▢ ▴◈◫▽◧◓▵,"- I'm not comfortable staying here, just us."


Из вывода ясно, что комбинация `- ` в начале строки не имеет связи с начальным символом в `dst`. удалим начало строки в `src`

In [35]:
# Удаление комбинации символов '- ' в начале строки в признаке 'src'
train_df['dst'] = train_df['dst'].str.lstrip('- ')

# Проверка результата
train_df.head()

,src,dst
0,◄▴◓◠▨ ◨▽◠▦◈◬◓▪▼◬▵,Intriguing.
1,▽◪◎◗▦◫▦◫ ▫▴▨◓◠◓ ▴▫◎◪▱◫ ◚▴ ◞◧▦◞▾▢▱◨▨ ◒◠◓◠◀▪▦◈◠▦...,He would need to repeat his vows in the land o...
2,◄▴◞◠▸▱◠◓▪◎◠ ◀◫▱◪ ▼◪◚◠▻ ◚▴◓▴◎◪◈◗▦ ◎◫?,You couldn't even answer my texts?
3,▯◪ ▨◠◈◠◓ ◞◭◓◠▫ ◳◠▻◬◳◧◓ ◞▴▦◗▦▨◫?,How fast do you go?
4,◈◠ ◧▱◠▦ ◀◫◓ ▨◠◉ ◂▱◠▽◈◠▦ ◀◠▷◞◪◈◗◳◧◓■ ◉◧◐▾▦▱◨◐▾ ...,"He's talking about a few right here in Lisbon,..."


### Поиск дубликатов

Проверим наличие дубликатов в обучающей части датасета:

In [39]:
duplicates = train_df[train_df.duplicated(keep=False)]
duplicates
        

,src,dst


Удалим все дубликаты, оставляя только первый вариант:

In [40]:
# Удаление дубликатов, оставляя только первый вариант
train_df = train_df.drop_duplicates(subset=['dst'], keep='last')

# Проверка результата
duplicates = train_df[train_df.duplicated(keep=False)]
duplicates

,src,dst


### Различаются по символам пунктуации

В языке зетан символы пунктуации похожи на наши, значит если символы используемые в `src` не совпадаю с `dst`, то это битый перевод:

In [51]:
import string

# Функция для нахождения всех символов пунктуации в тексте
def find_punctuation(text):
    punctuation_set = set()
    for char in text:
        if char in string.punctuation:
            punctuation_set.add(char)
    return punctuation_set

# Нахождение всех символов пунктуации в src и dst
src_punctuation = set()
dst_punctuation = set()

for split in dataset:
    for example in dataset[split]:
        src_punctuation.update(find_punctuation(example['src']))
        dst_punctuation.update(find_punctuation(example['dst']))

# Вывод разницы между множествами пунктуации
punctuation_difference = src_punctuation.symmetric_difference(dst_punctuation)
print(f"Символы пунктуации в src: {src_punctuation}")
print(f"Символы пунктуации в dst: {dst_punctuation}")
print(f"Разница между множествами пунктуации: {punctuation_difference}")

TypeError: 'NoneType' object is not iterable

### Поиск различий в количестве пробелов

In [46]:
# Функция для подсчета количества пробелов в строке
def count_spaces(text):
    return text.count(' ')

# Выбор строк, где разница в количестве пробелов больше одного
space_diff = train_df[abs(train_df['src'].apply(count_spaces) - train_df['dst'].apply(count_spaces)) > 4]

# Вывод найденных строк
space_diff

,src,dst
1,▽◪◎◗▦◫▦◫ ▫▴▨◓◠◓ ▴▫◎◪▱◫ ◚▴ ◞◧▦◞▾▢▱◨▨ ◒◠◓◠◀▪▦◈◠▦...,He would need to repeat his vows in the land o...
4,◈◠ ◧▱◠▦ ◀◫◓ ▨◠◉ ◂▱◠▽◈◠▦ ◀◠▷◞◪◈◗◳◧◓■ ◉◧◐▾▦▱◨◐▾ ...,"He's talking about a few right here in Lisbon,..."
5,◝▴▦ ◫◒▫◪▦ ▨◠◳▫◠◓◠▼◠▨ ◀◗◓◫ ◈▴◐◫▱◈◗◓▵,"For instance, Ben is not a guy who skips work."
20,◝◪▦ ◞◠◈▴▼▴ ◗◳◗ ◀◗◓ ◈◫▢◫ ◉▴▨▫◫◐◗◎◗▢◫ ◞◇◳▱◭▽◧◓▾◎▵,"I'm just saying, like, we're making a good sho..."
21,▢◠▫◪▦ ◀▴◈◪▱◫▦◫ ◇◈▩◳◧◓◨◎; ▻▴▨◗ ◧▦◠ ▦◪ ▽◠▻◠▼◠▨◞▪▦?,"And now I'm all taking care of, but what are y..."
...,...,...
299981,◞◠▦◠ ◠▽▦◬ ◒◪◳◫ ◞◧◓◠▼◠◐◬▢▵,I think that we would like to ask you the same...
299987,◌◓◠▦▨■ ◉▪▨ ◧◓▫◠▽◠▵,"Jimmy crack corn... - Frank, come on out!"
299993,▭◪▽■ ◧ ▴▻◪▽◈◫◓ ◀◨◓◠◳◠ ◕▴▱◎◗▽◂◓▵,"I... Hey, she hasn't been over here in a while."
299996,◝◓◳▼▴ ◉◂▼▾◐▾◎◠ ◫◳◫ ◀◠▨,"Brice, I need you take care of my child. I wil..."


### Дедубликация

## Подготовка данных

### Токенизатор

#### Нормализация, размер словаря

Нормализация включает в себя некоторую общую очистку, например, удаление ненужных пробельных символов, понижение регистра и/или удаление ударений. Посмотрим как выполняется нормализация токенизатором выбранной нами модели:

In [ ]:
example_num = 13

tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)

# Проверка нормализации текста
text = dataset['validation'][example_num]['src']
translate = dataset['validation'][example_num]['dst']

print("Модель: ", model_checkpoint)
print("Оригинальный текст:", text)
print("Нормализованный текст:", tokenizer.backend_tokenizer.normalizer.normalize_str(text))
print("Перевод:", translate)

# Проверка, является ли токенизатор быстрым
print("Токенизатор быстрый:", tokenizer.is_fast)

# Узнаем текущий размер словаря токенизатора
current_vocab_size = tokenizer.vocab_size
print(f"Текущий размер словаря токенизатора: {current_vocab_size}")


Судя по выводу текст подвергается минимальной нормализации, значит проблем с потерей символов быть не должно. Размер предобученного словаря токенизатора для выбранной модели равен 32100 токен. 

#### Размер контекстного окна

In [ ]:
# Получим максимальную длину последовательности (размер контекстного окна) модели 
max_length = tokenizer.model_max_length

print(f"Максимальная длина последовательности модели: {max_length}")

Максимальная длинна контекста для выбранной модели равна 512 токенов.

> ***Длина контекста (или контекстное окно)*** это максимальное количеству токенов, которые модель может обрабатывать одновременно.

#### Обучение нового токенизатора

Так как язык зетан является вымешленным (новым), необходимо обучить новый токенизатор:

In [ ]:
def batch_iterator(batch_size=10):
    for split in dataset:
        batch_size = batch_size // 2
        for i in range(0, len(dataset[split]), batch_size):
            src_list = dataset[split][i : i + batch_size]["src"]
            dst_list = dataset[split][i : i + batch_size]["dst"]
            yield [str(src) for src in src_list] + [str(dst) for dst in dst_list]


# vocab_size = 50000  # Можно начать с 32000 и экспериментировать

# Обучение токенизатора
new_tokenizer = tokenizer.train_new_from_iterator(batch_iterator(batch_size=1000), vocab_size=vocab_size)
print("Токенизатор обучен")

current_vocab_size = new_tokenizer.vocab_size
print(f"Текущий размер словаря токенизатора: {current_vocab_size}")


Проверим работу нового токенизатора на нескольких примерах исходного и целевого текста:

In [ ]:
example_num = 1

# Получаем строки из датасета
text = dataset['validation'][example_num]['src']
translate = dataset['validation'][example_num]['dst']

# Токенизируем текст и перевод, выводим результат
print(text)
print(new_tokenizer(text), end="\n\n")

print(translate)
print(new_tokenizer(translate), end="\n\n")

print(new_tokenizer.vocab)

Сохраним полученный новый токенизатор локально: 

In [ ]:
new_tokenizer.save_pretrained("./Model/new_tokenizer")

### Подготовка данных для обучения модели

Загрузим обученный ранее токенизатор:


In [ ]:
tokenizer = AutoTokenizer.from_pretrained("./Model/new_tokenizer")

Проверим токенизатор на паре предложений из тренировочной части датасета:

In [ ]:
dataset['train'][example_num]['src']

In [ ]:
example_num = 7
test_examples = [dataset['train'][example_num]['src'], dataset['train'][example_num]['dst']]
tokenizer(test_examples)

Для разных моделей может понадобиться разный подход в настройке перевода. Например для моделей семества `google-t5` необходим специальный преффикс. Реализуем это:

In [ ]:
print(model_checkpoint)

if model_checkpoint in ["google-t5/t5-small", "google-t5/t5-base", "google-t5/t5-larg", "google-t5/t5-3b", "google-t5/t5-11b", "artyomboyko/t5-small-finetuned-zetan-to-en"]:
    print("This is a T5 model")
    prefix = "translate Zetan to English: "
else:
    print("This is not a T5 model")
    prefix = ""

Прежде чем передать в модель тренировочнуй часть датасета необходимо его предварительно обработать. Напишем функцию которая подготовит датасет в которой:

- примеры из датасета будут подаваться на вход `tokenizer` с аргументом `truncation=True`. Это гарантирует, что входные данные, длина которых превышает ту, которую может обработать выбранная модель, будут усечены до максимальной длины, принимаемой моделью.

- Дополнение коротких последовательностей (Padding) будет реализованно позже (в коллаторе данных).

In [ ]:
# Установка максимальной длинны входной и выходной последовательности
max_input_length = 512  # max_length
max_target_length = 512 # max_length

source_lang = "zetan"
target_lang = "en"

def preprocess_function(examples):

    inputs = [prefix + ex for ex in examples['src']]
    targets = [ex for ex in examples['dst'] if ex is not None]

    # Токенизируем входные данные и цели
    model_inputs = tokenizer(inputs, max_length=max_input_length, truncation=True)

    with tokenizer.as_target_tokenizer():
        labels = tokenizer(targets, max_length=max_target_length, truncation=True)

    model_inputs["labels"] = labels["input_ids"]
    return model_inputs

Проверим функцию на нескольких примерах из датасета:

In [ ]:
preprocess_function(dataset['train'][:2])

Загрузим датасет без тестовой части, так как в ней нет целей. Увеличим валидационную часть датасета, и проведем токенизацию используя метод `map`.

Важно помнить, что увеличение размера валидационной части датасета может оказывать следующее влияние на процесс обучения модели:

***Положительное влияние:***
- Более точная оценка: Большая валидационная выборка обеспечивает более точную оценку производительности модели на невидимых данных. Это помогает лучше понять, насколько хорошо модель обобщает знания и избегает переобучения.
- Лучший выбор гиперпараметров: Больший объем данных для валидации позволяет более надежно сравнивать различные модели и выбирать оптимальные гиперпараметры, что в конечном итоге приводит к лучшей производительности.
- Раннее обнаружение переобучения: С увеличением размера валидационной выборки, переобучение (когда модель хорошо работает на тренировочных данных, но плохо на новых) становится заметнее на ранних стадиях, позволяя быстрее корректировать процесс обучения.

***Отрицательное влияние:***
- Меньше данных для обучения: Увеличение валидационной выборки означает уменьшение объема данных, доступных для обучения. Если общий объем данных ограничен, это может негативно сказаться на производительности модели.
- Увеличение времени обучения: Валидация на большем наборе данных занимает больше времени, что может замедлить процесс обучения.

Нужно найти баланс:
Ключевым моментом является нахождение правильного баланса между размером обучающей и валидационной выборок.
Общие рекомендации: Часто используется соотношение 80/20 или 70/30 для обучающей и валидационной выборок соответственно.



In [ ]:

# Стандартное разделение датасета.
data_files = {
    "train": "train.jsonl",
    "validation": "val.jsonl",
}

features = Features({
    'src': Value('string'),
    'dst': Value('string'),
})

dataset = load_dataset("json", data_files=data_files, features=features)
dataset



# # Загрудка датасета с ручным разделением на train и validation
# dataset_seed = 77 #  Уже использовал 42,
# dataset_test_size = 0.1 # Уже использовал 0.01

# features = Features({
#     'src': Value('string'),
#     'dst': Value('string'),
# })

# dataset = load_dataset('json', data_files=['train.jsonl', 'val.jsonl'], features=features)
# dataset = dataset['train'].train_test_split(test_size=0.01, seed=dataset_seed, shuffle=True)
# dataset['validation'] = dataset.pop('test')
# dataset 

In [ ]:
# Применение функции предобработки к датасету
tokenized_datasets = dataset.map(preprocess_function, batched=True, num_proc=20)

tokenized_datasets['train'][0]

# Обучение модели

## Загрузка контрольной точки модели

Загрузим необученную модель (модель со случайно сгенерированными весами):

In [ ]:
# Загрузка модели со случайными весами (если хотим обучать с нуля)
# model_checkpoint = "google-t5/t5-small"
# config = AutoConfig.from_pretrained(model_checkpoint)
# model = T5ForConditionalGeneration(config)

# # # Загрузка модели с предобученными весами с Hub для начала обучения
# model_checkpoint = "google-t5/t5-base"
model_checkpoint = "./Model/t5-small-finetuned-zetan-to-en/checkpoint-56256"
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)

# # Загрука модели с Hub уже предобученная мной (продолжение обучения)
# model_checkpoint = "artyomboyko/t5-small-finetuned-zetan-to-en"
# model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)
# tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)

print("Контрольная точка модели загружена.", model_checkpoint)

# print(f"Текущий размер словаря модели {model.config.vocab_size}.")
# model.resize_token_embeddings(vocab_size)
# print(f"Новый размер словаря модели {model.config.vocab_size}.")




## Определение метрики

Определим метрику которая будет использоваться для определения качества предсказаний модели в процессе обучения и несколько функций для предобработки результатов:

In [ ]:
# Загрузим метрику
metric = load("sacrebleu")

# Функция для постобработки текста
def postprocess_text(preds, labels):
    preds = [pred.strip() for pred in preds]
    labels = [[label.strip()] for label in labels]

    return preds, labels

# Функция для вычисления метрик
def compute_metrics(eval_preds):
    preds, labels = eval_preds
    if isinstance(preds, tuple):
        preds = preds[0]
    decoded_preds = tokenizer.batch_decode(preds, skip_special_tokens=True)

    # Замените -100 в метках, так как мы не можем их декодировать.
    labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
    decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)

    # Немного простой постобработки
    decoded_preds, decoded_labels = postprocess_text(decoded_preds, decoded_labels)

    result = metric.compute(predictions=decoded_preds, references=decoded_labels)
    result = {"bleu": result["score"]}

    prediction_lens = [np.count_nonzero(pred != tokenizer.pad_token_id) for pred in preds]
    result["gen_len"] = np.mean(prediction_lens)
    result = {k: round(v, 4) for k, v in result.items()}
    return result

## Определение аргументов обучения

Определим основные настройки обучения:

При настройках:
batch_size = 112    
lr = 1e-3    
dataset = dataset['train'].train_test_split(test_size=0.01, seed=dataset_seed, shuffle=True)    

Хватит 45 эпох чтобы получить BLEU:    

Лучшая метрика: 18.1667    
Контрольная точка лучшей модели:    

In [ ]:
batch_size = 64 # Оптимальное на текущий момент 64! Изначально было 16, если в VRAM помещается больше, можно увеличить. Это может ускорить обучение и улучшить качество модели.
epochs = 50 # Для подбора lr нужно брать минимум 6 эпох, чтобы стабилизировался лосс и метрика
lr = 1.5e-5 # Оптимальный на текущий момент от 1e-3 для T5-small (подбором 2e-3). T5-base LR 1e-3 (BLEU 3,00)

model_name = model_checkpoint.split("/")[-1]
args = Seq2SeqTrainingArguments(
    output_dir=f"./Model/{model_name}-finetuned-{source_lang}-to-{target_lang}",
    eval_strategy = "epoch",                # Оценка модели в конце каждой эпохи (возможные варианты: "no", "steps", "epoch")
    save_strategy = "epoch",                # Сохранение модели в конце каждой эпохи (возможные варианты: "no", "steps", "epoch")
    learning_rate=lr,
    per_device_train_batch_size=batch_size,
    per_device_eval_batch_size=batch_size,
    weight_decay=0.03,                         # интенсивность L2-регуляризации (изначально было 0.01, при 0.05 - BLEU 3,52) Для T5-base тоже пробовал 0.05 - BLEU .
    save_total_limit=3,                     # Максимальное количество точек сохранения моделей (одновременно). Чтобы не забивать диск сохранениями на каждой эпохе (или через опеределенное количество шагов)
    num_train_epochs=epochs,
    predict_with_generate=True,             # Использование generate для предсказания
    fp16=True,                                                
    push_to_hub=True,                       # Публиковать модель на Hugging Face Hub
    hub_private_repo = True,                # Использовать для хранения модели приватный репозиторий
    optim = "adamw_torch",                  # Оптимизатор
    load_best_model_at_end=True,            # Загрузка лучшей модели в конце обучения
    metric_for_best_model="bleu",           # Метрика для выбора лучшей модели
    greater_is_better = True,               # Лучшая модель - с наибольшим значением метрики
    report_to=["tensorboard"],              # Отправка метрик в TensorBoard
    warmup_steps=250,                       # Количество шагов для warmup
    lr_scheduler_type="linear",             # Линейное уменьшение learning rate
    save_safetensors=False,     
    gradient_accumulation_steps = 4,        # Пробуем 2, 
)

## Определение коллатора данных

DataCollator используется для подготовки батчей данных перед подачей их в модель. Основные задачи, которые выполняет DataCollator:
- Создание батчей: Он группирует отдельные примеры данных в батчи, что повышает эффективность обучения и инференса.
- Обработка данных: Может выполнять предварительную обработку данных, например:
- Паддинг: Добавление специальных токенов (padding tokens) к последовательностям разной длины, чтобы сделать их одинаковыми. Это необходимо, так как модели обычно требуют входные данные фиксированной длины.
- Маскирование: Игнорирование определенных токенов во время обучения, например, при реализации masked language modeling.
- Создание дополнительных входных данных: Например, добавление меток классов или идентификаторов.

***Типы DataCollator:***

В Transformers есть несколько предопределенных DataCollator для разных задач:
- `DataCollatorWithPadding`: Динамически добавляет паддинг к последовательностям в батче.
- `DataCollatorForLanguageModeling`: Используется для задач языкового моделирования, применяет маскирование к входным последовательностям.
- `DataCollatorForSeq2Seq`: Для seq2seq задач, таких как машинный перевод.
- `DefaultDataCollator`: Просто объединяет данные в батчи без дополнительной обработки.

Для нашей задачи (перевод) нам нужен специальный коллатор данных (`DataCollatorForSeq2Seq`), который дополнит входные данные и метки до максимальной длины в батче:

In [ ]:
data_collator = DataCollatorForSeq2Seq(tokenizer, model=model)

## Обучение модели

Затем нам нужно просто передать все это вместе с нашими датасетами в `Seq2SeqTrainer`:

In [ ]:
trainer = Seq2SeqTrainer(
    model,
    args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["validation"],
    data_collator=data_collator,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

Теперь мы можем дообучить нашу модель, просто вызвав метод `train`:

In [ ]:
trainer.train()


Посмотрим на лучшу метрику и лучшую контрольную точку полученные во время обучения:

In [ ]:
trainer.state.best_metric

In [ ]:
trainer.state.best_model_checkpoint

In [ ]:
print(f"Лучшая метрика: {trainer.state.best_metric}")
print(f"Контрольная точка лучшей модели: {trainer.state.best_model_checkpoint}")

После обучения модели будет автоматически загружена лучшая из моделей полученных в процессе обучения, сохраним ее и токенизатор на Hub:

In [ ]:
trainer.push_to_hub()
# model = AutoModel.from_pretrained('./Model/t5-small-finetuned-zetan-to-en/checkpoint-135507')
# model.push_to_hub(repo_id=f"{model_name}-finetuned-{source_lang}-to-{target_lang}")
tokenizer.push_to_hub(repo_id=f"{model_name}-finetuned-{source_lang}-to-{target_lang}")

In [ ]:
model = f

# Подготовка файла с ответами

Загрузим модель и токенизатор с Hub:

In [ ]:
import torch
import numpy as np

from transformers import AutoModel, AutoConfig, AutoTokenizer, AutoModelForSeq2SeqLM, T5ForConditionalGeneration, Seq2SeqTrainingArguments, DataCollatorForSeq2Seq, Seq2SeqTrainer
from datasets import load_dataset, Features, Value, concatenate_datasets, Dataset, DatasetDict
from huggingface_hub import notebook_login
from evaluate import load

device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
print('{} device is available'.format(device))

seed = 71

if torch.cuda.is_available() and device.type == 'cuda':
    np.random.seed(seed)
     # Если используете CUDA, устанавливаем seed для GPU                
    torch.cuda.manual_seed_all(seed)
    # Для полной воспроизводимости также можно зафиксировать алгоритмы cuDNN
    torch.backends.cudnn.deterministic = True   
    torch.backends.cudnn.benchmark = False
else:
    # Устанавливаем seed для CPU генератора случайных чисел PyTorch
    np.random.seed(seed)
    torch.manual_seed(seed)


# Несколько оптимизаций которые могут ускорить процесс обучения
if torch.backends.cudnn.is_available():
    print(f'Using cudnn version: {torch.backends.cudnn.version()}')
    if device.type == 'cuda':
        torch.backends.cudnn.enabled = True
        torch.backends.cudnn.benchmark = True

In [ ]:
notebook_login() # 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

In [ ]:
# # translator_checkpoint = "artyomboyko/t5-small-finetuned-zetan-to-en"
translator_best_checkpoint = "./Model/t5-small-finetuned-zetan-to-en/checkpoint-155915"
model_checkpoint = "artyomboyko/t5-small-finetuned-zetan-to-en"
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)
model.to(device)

Формат решения: json-lines в формате, аналогичной валидации, с переводами исходных текстов в тестовой выборке

Загрузим тестовую часть датасета:

In [ ]:
data_files = {
    "test": "test_no_reference.jsonl"
}

features = Features({
    'src': Value('string'),
    'dst': Value('string'),
})

test_dataset = load_dataset("json", data_files=data_files, features=features)

print(test_dataset['test'][0])

Для формирования файла ответов необходимо сгенерировать перевод для всех признаков `src` в тестовой части датасета и экспортировать результат в файл json line. Сделаем это следующим образов. Создадим функцию, которая будет принимать на вход пример из датасета и возвращать его перевод. Применим функцию к датасету с помщью функции `map` библиотеки Datasets, после чего сохраним файл:

In [ ]:
def translate_example(example, prefix="translate Zetan to English: "):
    input_text = example["src"]
    input_ids = tokenizer(prefix + input_text, return_tensors="pt").input_ids.to(device)
    outputs = model.generate(input_ids)    
    translated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
    example["dst"] = translated_text
    return example

# Применение функции перевода к тестовому датасету
submission = test_dataset.map(translate_example)

# Сохранение результатов в файл
submission['test'].to_json('submission.jsonl', force_ascii=False)

# Дополнительные материалы

По основной задаче:

- [Training your own tokenizer from scratch](https://github.com/huggingface/notebooks/blob/main/examples/tokenizer_training.ipynb)
- [Translation fine-tuning example](https://github.com/huggingface/notebooks/blob/main/examples/translation.ipynb)

По проблемам с контейнером:

- [Jupyter notebook executions turn grey in Visual studio code](https://stackoverflow.com/questions/73481249/jupyter-notebook-executions-turn-grey-in-visual-studio-code)
- [Jupyter Cache](https://jupyter-cache.readthedocs.io/en/latest/index.html)

# Тестовая часть

Используется для отладки кода и проведения экспериментов.